packages/components/src/internal/components/domainproperties/DomainForm.tsx

```diff
@@ -388,28 +427,149 @@ export class DomainFormImpl extends React.PureComponent<IDomainFormInput, IDomai
         } else this.applyAddField();
     };
 
+    toggleSelectAll = () => {
+        const { domain } = this.props;
+        const { selectAll, visibleSelection } = this.state;
+
+        let newVisibleSelection = new Set([...visibleSelection]);
+        const toggledFields = domain.fields.map((field, index) => {
+            if (field.visible) {
+                newVisibleSelection = applySetOperation(newVisibleSelection, index, !selectAll);
+                return field.set('selected', !selectAll);
+            } else {
+                return field;
+            }
+        });
+
+        const updatedDomain = domain.merge({
+            fields: toggledFields,
+        }) as DomainDesign;
+        this.onDomainChange(updatedDomain, true);
+        this.setState((state) => ({selectAll: !state.selectAll, visibleSelection: newVisibleSelection}));
+    };
+
+    clearAllSelection = () => {
+        const { domain } = this.props;
+        const fields = domain.fields.map((field) => {
+            return field.set('selected', false);
+        });
+        const updatedDomain = domain.merge({fields}) as DomainDesign;
+        this.onDomainChange(updatedDomain, true);
+        this.setState({selectAll: false, visibleSelection: new Set()});
+    }
+
     onExportFields = () => {
         const { domain } = this.props;
+        const { visibleSelection } = this.state;
         const fields = domain.fields;
-        const filteredFields = fields.filter((field: DomainField) => field.visible);
+        let filteredFields = fields.filter((field: DomainField) => field.visible);
+        // Respect selection, if any selection exists
+        filteredFields = visibleSelection.size > 0
+            ? filteredFields.filter((field: DomainField) => field.selected)
+            : filteredFields;
+
         const fieldData = filteredFields.map(field => DomainField.serialize(field, false)).toArray();
         const fieldsJson = JSON.stringify(fieldData, null, 4);
 
         downloadJsonFile(fieldsJson, generateNameWithTimestamp('Fields') + '.fields.json');
     };
 
+    renderBulkFieldDeleteConfirm = () => {
+        const { domain } = this.props;
+        const { bulkDeleteConfirmInfo } = this.state;
+        const undeletableNames = bulkDeleteConfirmInfo.undeletableFields.map((i) => {return domain.fields.get(i).name});
+        const { howManyDeleted, undeletableWarning } = generateBulkDeleteWarning(bulkDeleteConfirmInfo, undeletableNames);
+
+        const thisFieldPlural = bulkDeleteConfirmInfo.deletableSelectedFields.length > 1 ? 'these fields' : 'this field';
```

Comment:
Seen in quick manual testing pass, if I only select non-deletable fields and then click Delete, the dialog says 0 of N fields will be deleted. That part is right, but the rest of the text in the dialog and the buttons seems odd in this case. Maybe we instead want to detect this case and open a simple modal that says "None of the selected fields can be deleted" with a single "close" button?

Reply:
Ah makes sense, done! I have a title of 'Cannot Delete Required Fields' and modal body text of 'None of the selected fields can be deleted.' I could run this by the design groupchat?

Reply:
That looks good

Reply:
Actually on second thought, the word "required" in this context refers to the required checkbox column. I'd suggest changing the title of this modal to "Cannot Delete Locked Fields" or to keep the modal title consistent for both cases (i.e. "Confirm Delete Selected Fields").